feat(expo-cli): extract android credentials flow

wkozyra95 opened the pull request:
# Why

Part of the turtle-v2 support

# How
- extract android credentials flow from AndroidBuilder
- improve error messages in `expo credentials:manager`
- remove all uses of xdl/src/credentials/Credentials.ts` for android

# Test plan
`fetch:android:***` with valid credentials, invalid credentials, no credentials
`build:android` with  and without credentials
`build:android --non-interactive` with and without credentials
`build:android --clear-credentials --non-interactive` with and without credentials
`build:android --kesytore-path *** --keystore-alias **`

dsokal commented on packages/expo-cli/src/credentials/views/SetupAndroidKeystore.ts:
```diff
@@ -0,0 +1,18 @@
+  constructor(private experienceName: string) {}
+
+  async open(ctx: Context): Promise<IView | null> {
+    const { keystore, keystorePassword, keyAlias, keyPassword } =
+      (await ctx.android.fetchKeystore(this.experienceName)) ?? {};
+    if (!keystore || !keystorePassword || !keyAlias || !keyPassword) {
+      return new UpdateKeystore(this.experienceName);
+    }
+    return null;
```
```suggestion
    const keystore = await ctx.android.fetchKeystore(this.experienceName);
    if (!keystore) {
      return new UpdateKeystore(this.experienceName);
    } else {
      return null;
    }
```

dsokal replied:
Destructuring the result of`ctx.android.fetchKeystore` doesn't make sense. 

The result is either null or Keystore:
`public async fetchKeystore(experienceName: string): Promise<Keystore | null> {`

and this is the Keystore type definition:
```
type Keystore = {
    keystore: string;
    keystorePassword: string;
    keyPassword: string;
    keyAlias: string;
};
```

wkozyra95 replied:
In case of invalid data in db

dsokal commented on packages/expo-cli/src/credentials/actions/promptForCredentials.ts:
```diff
@@ -28,6 +28,11 @@ export type CredentialSchema<T> = {
+  provideMethodQuestion?: {
+    question?: string;
+    expoGenerated?: string;
+    manually?: string;
```
```suggestion
    userProvided?: string;
```

dsokal commented on packages/expo-cli/src/credentials/actions/promptForCredentials.ts:
```diff
@@ -61,14 +66,20 @@ export async function getCredentialsFromUser<T extends Results>(
+        value: false,
+      },
+      {
+        name: schema?.provideMethodQuestion?.manually ?? 'I want to upload my own file',
```
```suggestion
        name: schema?.provideMethodQuestion?.userProvided ?? 'I want to upload my own file',
```

dsokal commented on packages/expo-cli/src/credentials/api/android.ts:
```diff
@@ -0,0 +1,72 @@
+    }
+  }
+
+  private async _ensureCredntials(experienceName: string): Promise<void> {
```
```suggestion
  private async _ensureCredentialsFetched(experienceName: string): Promise<void> {
```

dsokal commented on packages/expo-cli/src/credentials/api/android.ts:
```diff
@@ -0,0 +1,72 @@
+
+  public async fetchKeystore(experienceName: string): Promise<Keystore | null> {
+    await this._ensureCredntials(experienceName);
+    return this.credentials[experienceName]?.keystore;
```
Hmmm, I think there is a type mismatch. The return type for this function is `Keystore | null` but I think in the case where the credentials for a given `experienceName` don't exist - the result will be `undefined`.

dsokal commented on packages/expo-cli/src/credentials/context.ts:
```diff
@@ -41,6 +43,9 @@ export class Context {
     return this._apiClient as ApiV2;
   }
+  get android(): AndroidApi {
+    return this._androidApiClient as AndroidApi;
```
Hmmm, I don't like this `as AndroidApi`. Maybe it's worth adding a conditional to check if `this._androidApiClient` is undefined? And then throw some error if it's undefined.

dsokal replied:
Yes, I know we don't have such conditionals in other getters in this class.

wkozyra95 replied:
those getters are only unassigned if the context was not initialized

wkozyra95 replied:
I added api client when called outside the project context

dsokal commented on packages/expo-cli/src/credentials/views/AndroidKeystore.ts:
```diff
@@ -0,0 +1,231 @@
+  if (!keystorePath || !keyAlias || !keystorePassword || !keyPassword) {
+    console.log(keystorePath, keyAlias, keystorePassword, keyPassword);
+    throw new Error(
+      'In order to provide a Keystore through the CLI parameters, you have to pass --keystore-alias parameter, --keystore-path parameter and set EXPO_ANDROID_KEY_PASSWORD and EXPO_ANDROID_KEYSTORE_PASSWORD environment variables.'
```
```suggestion
      'In order to provide a Keystore through the CLI parameters, you have to pass --keystore-alias, --keystore-path parameters and set EXPO_ANDROID_KEY_PASSWORD and EXPO_ANDROID_KEYSTORE_PASSWORD environment variables.'
```

wkozyra95 commented on packages/expo-cli/src/commands/build/credentials/prepareAndroid.ts:
```diff
@@ -0,0 +1,51 @@
+  clearCredentials?: boolean;
+}
+
+async function prepareBuildCredentials(projectDir: string, options: Options): Promise<Keystore> {
```
`Are there any other credentials than build credentials?` - push credentials

wkozyra95 commented on packages/expo-cli/src/commands/build/credentials/prepareAndroid.ts:
```diff
@@ -0,0 +1,51 @@
+    await runCredentialsManager(ctx, new RemoveKeystore(experienceName));
+  }
+
+  const paramKeystore = await getKeystoreFromParams(options);
+  if (paramKeystore) {
+    await useKeystore(ctx, experienceName, paramKeystore);
+  } else {
+    const view = new SetupAndroidKeystore(experienceName);
+    await runCredentialsManager(ctx, view);
+  }
```
I think that credentials manager shouldn't be aware of cli options. It makes reusing that code a lot more complicated

wkozyra95 commented on packages/expo-cli/src/commands/build/credentials/prepareAndroid.ts:
```diff
@@ -0,0 +1,51 @@
+  }
+  const keystore = await ctx.android.fetchKeystore(experienceName);
+  if (!keystore) {
+    throw new Error("No keystore assigned for this app, this shouldn't happen");
```
fetchKeystore can return null, but not here because we just ensured that there is keystore

dsokal commented on packages/expo-cli/src/credentials/views/AndroidKeystore.ts:
```diff
@@ -0,0 +1,231 @@
+      };
+    } catch (error) {
+      log.warn(
+        "If you don't provide your own Android keystore, it will be generated on our servers during the next build"
```
This message is a little weird. If I understand the code in `try` correctly - the error can be thrown if we failed to generate a keystore. For example, if the `keytool` command had failed.

wkozyra95 commented on packages/expo-cli/src/commands/build/credentials/android.ts:
```diff
@@ -0,0 +1,90 @@
+  }
+}
+
+async function collectAndValidateCredentials(
```
Initially, I wanted to implement one credentials flow, but after a discussion about generating Keystore  server-side I had to separate both logic
`tryCollect..` was supposed to be used by turtle-v1 builds and `collectAndVal....` in turtle-v2

but  after going through what we discussed turtle-sync this behavior can't be reused, so I need to change it either way

wkozyra95 commented on packages/expo-cli/src/commands/build/credentials/android.ts:
```diff
@@ -0,0 +1,90 @@
+  if (paramKeystore) {
+    await useKeystore(ctx, experienceName, paramKeystore);
+  } else if (await keytoolCommandExists()) {
+    const view = new SetupAndroidKeystore(experienceName);
+    await runCredentialsManager(ctx, view);
```
I prefer to detect if keytool is available rather than falling back after failed generation. Unless you are suggesting both.

dsokal commented on packages/expo-cli/src/commands/build/AndroidBuilder.ts:
```diff
@@ -53,207 +56,42 @@ See https://docs.expo.io/distribution/building-standalone-apps/#2-configure-appj
+      try {
+        await runCredentialsManager(ctx, new SetupAndroidKeystore(experienceName));
+      } catch (_) {
+        /* */
```
Maybe add some comment here

dsokal commented on packages/expo-cli/src/credentials/views/AndroidKeystore.ts:
```diff
@@ -0,0 +1,234 @@
+  if (!keystorePath || !keyAlias || !keystorePassword || !keyPassword) {
+    console.log(keystorePath, keyAlias, keystorePassword, keyPassword);
+    throw new Error(
+      'In order to provide a Keystore through the CLI parameters, you have to pass --keystore-alias, --keystore-path parameters and set EXPO_ANDROID_KEY_PASSWORD and EXPO_ANDROID_KEYSTORE_PASSWORD environment variables.'
```
Hmmm, we need to be consistent about the capitalization of `Keystore`. In some places, it's not capitalized and in some places it is.

quinlanj commented on packages/expo-cli/src/credentials/credentials.ts:
```diff
@@ -121,15 +121,22 @@ export type FcmCredentials = {
 export const keystoreSchema: CredentialSchema<Android.Keystore> = {
   id: 'keystore',
   name: 'Android Keystore',
+  provideMethodQuestion: {
```
Is there a reason why the keystore generation prompts are written this way instead of the usual 
```
Will you provide your own <thing>?
- Let Expo handle the process
- I want to upload my own file
```
It would simplify the code if we just used the normal prompts which accomplish the same thing, unless there is good reason to use custom prompts

wkozyra95 replied:
A lot of people don't understand what it means to "let expo handle it" and assume that there is a way to automatically generate new keystore that will work when uploading what often leads to need to reset keystore via google support.

wkozyra95 commented on packages/expo-cli/src/commands/build/AndroidBuilder.ts:
```diff
@@ -53,207 +56,43 @@ See https://docs.expo.io/distribution/building-standalone-apps/#2-configure-appj
 
-  platform(): Platform {
-    return ANDROID;
+async function keytoolCommandExists(): Promise<boolean> {
```
the only issue with compatibility I found is that on some version store type defaulted to PKCS instead JKS, but it was fixed some time ago. If local generation of keystore fails we are not stoping build command.

wkozyra95 replied:
> If local generation of keystore fails we are not stoping build command.

I removed that behaviour because it complicated code significantly, but I don't think this will be a problem. It would be very weird case for keytool to exist in PATH but fail to generate keystore

wkozyra95 commented on packages/expo-cli/src/commands/build/AndroidBuilder.ts:
```diff
@@ -53,207 +56,43 @@ See https://docs.expo.io/distribution/building-standalone-apps/#2-configure-appj
+      }
+    } else {
+      log.warn(
+        'Program keytool was not found in PATH, new Keystore will be generated on Expo servers'
```
keytool is not specific to android I'll change that to `The `keytool` utility was not found in your PATH`

dsokal commented on packages/expo-cli/src/credentials/views/SetupAndroidKeystore.ts:
```diff
@@ -0,0 +1,51 @@
+    if (this.options.nonInteractive) {
+      if (this.options.allowMissingKeystore) {
+        log.warn(
+          'There is no Keystore assigned for this, new one will be generated on Expo servers.'
```
```suggestion
          'There is no Keystore assigned for this ??PUT_STH_HERE??, a new one will be generated on Expo servers.'
```

wkozyra95 commented:
@quinlanj thanks for pointing `nonInteractive` out, I was not aware that this flag was added. I added fixes for non-interactive mode. It finishes successfully only when Keystore is already setup or there is no Keystore(generates on the worker). I don't want to either create Keystore or update in the nonInteractive mode because those operations are too dangerous to be run on CI.

I'm not sure about `--generate-keystore` flag, Keystore will always be generated on the worker if it's missing, so it's the default behaviour and setting this flag to false also does not make sense because the build will fail. I left the flag there, but I'm not checking it anywhere

As for testing, I added AndroidApi test and I'm still working on views tests that will be added in separate PR.

quinlanj commented:
@wkozyra95 sounds good, thanks for writing the tests! If `--generate-keystore` doesn't make sense to have in the current and future implementation, let's add a warning or smth to tell ppl its deprecated

```
         log.warning(
            'This flag is deprecated and does not do anything. A Keystore will always be generated on the Expo servers if it's missing.'
          );
```

quinlanj commented on packages/expo-cli/src/commands/build/credentials/android.ts:
```diff
@@ -0,0 +1,90 @@
+    await runCredentialsManager(ctx, view);
+  } else {
+    log.warn(
+      'Program keytool was not found in PATH, new keystore will be generated by Expo servers'
```
Either works, but I prefer `on`!
